In [2]:
from keras.applications import VGG16

In [3]:
vgg_cnn = VGG16()

553467096/553467096 [==============================] - 91s 0us/step


In [5]:
print(vgg_cnn.summary())

Model: "vgg16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 224, 224, 3)]     0         
                                                                 
 block1_conv1 (Conv2D)       (None, 224, 224, 64)      1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 224, 224, 64)      36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 112, 112, 64)      0         
                                                                 
 block2_conv1 (Conv2D)       (None, 112, 112, 128)     73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 112, 112, 128)     147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 56, 56, 128)       0     

**Size of kernels in first conv2d layer?** <br>
Let kernel size is n, then: <br>
64*n*n*channels + bias = total params <br>
64*n^2*3 + 64 = 1792 <br>
64*3*n^2 = 1728 <br>
$\implies$ n = 3 <br>
So, size of kernels in first conv2d layer is 3x3

In the architecture, you will see that flatten layer has vector size 25088 <br>
But, dense layer connected to it has 4096 units <br>
Dimensionality reduction of fearures is done here.
<hr>

VGG is trained on general images of 1000 categories. <br>
To adust the classes number for our use we have two choices:
- use classes parameter to ajust the output layer
- set include_top parameter to false to remove the entire head, then connect your custom head.

Sometimes we would like to keep the weights of pre-trained model. Sometimes we don't want to do that and train the model completely on our dataset. These depends on our goals. <br>
VGG is pre-trained to general images which are visible around us. If our goal is to do classification on medical images then using the pre-learned weights og VGG is not a good idea. We can set the weight parameter to none to discard this. <br> <br>
Beware: If you are taking the pre-learned weights then you have to freeze them.

Lets use it for Monkey class classification

In [6]:

vgg_cnn_md = VGG16(include_top=False,weights=None,input_shape=(400,300,3))

In [7]:
vgg_cnn_md.summary()

Model: "vgg16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_2 (InputLayer)        [(None, 400, 300, 3)]     0         
                                                                 
 block1_conv1 (Conv2D)       (None, 400, 300, 64)      1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 400, 300, 64)      36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 200, 150, 64)      0         
                                                                 
 block2_conv1 (Conv2D)       (None, 200, 150, 128)     73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 200, 150, 128)     147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 100, 75, 128)      0     

In [10]:
from keras.models import Model
from keras.layers import Flatten
from keras.layers import Dense

In [11]:
def create_custom_vgg16():
    flattened_image = Flatten()(vgg_cnn_md.output)
    fc1_output = Dense(units=512,activation="relu")(flattened_image)
    nw_output = Dense(units=10,activation="softmax")(fc1_output)
    
    our_custom_cnn = Model(inputs=vgg_cnn_md.input, outputs=nw_output)
    return our_custom_cnn

In [12]:
our_custom_vgg16_cnn = create_custom_vgg16()

In [13]:
our_custom_vgg16_cnn.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_2 (InputLayer)        [(None, 400, 300, 3)]     0         
                                                                 
 block1_conv1 (Conv2D)       (None, 400, 300, 64)      1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 400, 300, 64)      36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 200, 150, 64)      0         
                                                                 
 block2_conv1 (Conv2D)       (None, 200, 150, 128)     73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 200, 150, 128)     147584    
                                                                 
 block2_pool (MaxPooling2D)  (None, 100, 75, 128)      0     